### PR TITLE
Enable Merkle tree-based replication

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

This PR does a few things:
1. Enable new storage format and protocol by default
2. Fix handling of system collections so they can use the new protocol
3. Add some plumbing to upgrade old collections to the new format. Currently only works in singleserver mode. Cluster upgrade will come in a separate PR.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as `replication_sync`.

This PR adds tests that were used to verify the upgrade changes. In particular, updated `upgrade_data_*` tests.

Jenkins: http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9272/ (all blue except known unrelated failures)

### CODE REVIEW

File: arangod/Network/Methods.h [hunk 1]
```diff
@@ -96,6 +96,14 @@ FutureRes sendRequest(ConnectionPool* pool, DestinationId destination,
                       RequestOptions const& options = {},
                       Headers headers = {});
 
+/// @brief send a request to all responsible servers for a given shard; use this
+/// method to send a message to followers as well as leader
+std::vector<FutureRes> sendRequests(ConnectionPool* pool, DestinationId destination,
```
Comment (graetzer) [inline arangod/Network/Methods.h:101]: seems like this should not be here ? since it really just deals with shards the behaviour is also not that related to the rest of the sendRequest methods

Comment (danielhlarkin) [reply arangod/Network/Methods.h:101]: I suppose. I mean ... the purpose is to do exactly what `sendRequest` does, just with multiple targets. We could potentially unpack this so that it receives a vector of resolved destinations, and then use that as the internal basis for `sendRequest` (with a single vector entry). It just seems odd to me to reproduce the `sendRequest` functionality outside the `network` space just to get an overload on the destination resolution behavior.

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9282/ (all blue except a cluster timeout)
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-linux/11017/ (all blue)

**Comment by graetzer** [comment]
question: why do we do this new objectId / old ObjectId business? can we not just rewrite this stuff on the fly

**Comment by graetzer** [comment]
revisionIds and LocalDocumentIds are based on the same atomic counter, no ?

File: arangod/Network/Methods.h (hunk 1 above)
Comment (graetzer) [reply arangod/Network/Methods.h:101]: but its not the same thing, it just works for `shard:` destinations

**Comment by danielhlarkin** [comment]
@graetzer It's subtle. The `_rev` values should all be unique within a given collection (ignoring for now smart edge collections) since they are HLC values. Same goes for `LocalDocumentId` values. But the two sets of identifiers can be generated on different machines without proper communication to ensure uniqueness between, so we could possibly have a `LocalDocumentId` with the same value as a `_rev` from a different document.

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9308/ (all blue)

File: arangod/RocksDBEngine/RocksDBIndex.cpp [hunk 2]
```diff
@@ -327,3 +332,38 @@ RocksDBKeyBounds RocksDBIndex::getBounds(Index::IndexType type, uint64_t objectI
...
+    TRI_ASSERT(_tempObjectId == 0 || plannedTempObjectId == 0);
+    // just temp id has changed
+    std::uint64_t oldId = _tempObjectId;
+    _tempObjectId = plannedTempObjectId;
+    if (oldId != 0) {
+      try {
+        RocksDBKeyBounds bounds = getBounds(type(), oldId, unique());
+        return rocksutils::removeLargeRange(engine.db(), bounds, true, true);
+      } catch (arangodb::basics::Exception& ex) {
+        if (ex.code() != TRI_ERROR_NOT_IMPLEMENTED) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBIndex.cpp:353]: I guess we are catching only this type of error because `RocksDBIndex::getBounds` can throw it.
But it should only throw it for unknown index types, so I am wondering if we need this special handling here. Or is there a specific reason why we are ignoring this particular error. If so, a code comment would be useful.

File: tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+          RETURN doc`;
+    const geoExplain = db._createStatement(geoQuery).explain({});
+    assertNotEqual(-1, geoExplain.plan.rules.indexOf('geo-index-optimizer'));
+    const geoResults = db._query(geoQuery).toArray();
+    assertEqual(geoResults.length, 3);
+    assertEqual(geoResults[0].num, 1);
+    assertEqual(geoResults[1].num, 2);
+    assertEqual(geoResults[2].num, 3);
+
+    // non-unique skiplist
```
Comment (jsteemann) [inline tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js:121]:
```suggestion
    // fulltext index
```

File: arangod/VocBase/Methods/Collections.cpp [hunk 4]
```diff
@@ -892,6 +893,30 @@ futures::Future<Result> Collections::warmup(TRI_vocbase_t& vocbase,
   return futures::makeFuture(res);
 }
 
+futures::Future<Result> Collections::upgrade(TRI_vocbase_t& vocbase,
+                                             LogicalCollection const& coll) {
+  ExecContext const& exec = ExecContext::current();  // disallow expensive ops
+  if (!exec.canUseCollection(coll.name(), auth::Level::RO)) {
```
Comment (jsteemann) [inline arangod/VocBase/Methods/Collections.cpp:899]: Is read-only access sufficient for performing the upgrade? Or should we require read-write permissions? Not sure if the `RO` is intentional here or not. If so, a code comment would help.

File: arangod/VocBase/LogicalCollection.cpp [hunk 5]
```diff
@@ -938,6 +949,16 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice, b
...
   _waitForSync = Helper::getBooleanValue(slice, "waitForSync", _waitForSync);
   _sharding->setWriteConcernAndReplicationFactor(writeConcern, replicationFactor);
+  bool useRevs = Helper::getBooleanValue(slice, StaticStrings::UsesRevisionsAsDocumentIds,
+                                         _usesRevisionsAsDocumentIds);
+  if (!useRevs && _usesRevisionsAsDocumentIds) {
+    return Result(
+        TRI_ERROR_BAD_PARAMETER,
+        "collection uses revisions as document IDs, cannot set this to false");
+  }
+  _usesRevisionsAsDocumentIds = useRevs;
```
Comment (jsteemann) [inline arangod/VocBase/LogicalCollection.cpp:959]: Is it possible to set the `useRevisionsAsDocumentIds` from the outside, via a call to `db.<collection>.properties({ useRevisionsAsDocumentIds: true })`?
Is this harmful if this is done (un)intentionally on a collection that still uses the old format? Doesn't the collection need to be upgraded first?

Or, put differently: should we expose this in the (also) end-user facing collection properties API?

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 6]
```diff
@@ -614,6 +616,35 @@ Result RocksDBMetaCollection::bufferTruncate(rocksdb::SequenceNumber seq) {
...
+    std::uint64_t oldId = (plannedTempObjectId == 0) ? _tempObjectId : 0;
+    _tempObjectId = plannedTempObjectId;
+    if (oldId != 0) {  // need to clean up the old range
+      RocksDBKeyBounds bounds = RocksDBKeyBounds::CollectionDocuments(oldId);
+      res = rocksutils::removeLargeRange(engine.db(), bounds, true, true);
+    }
+  } else if (plannedTempObjectId != _tempObjectId) {
+    TRI_ASSERT(plannedObjectId != _objectId);
+    TRI_ASSERT(plannedObjectId != 0);
+    TRI_ASSERT(plannedObjectId = _tempObjectId);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:637]: I think this in untentional, as the assert will have side-effects apart from raising an assertion failure!

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 7]
```diff
@@ -69,16 +73,499 @@
...
+    builder.add(StaticStrings::UsesRevisionsAsDocumentIds,
+                arangodb::velocypack::Value(true));
+  } else {
+    builder.add(key);
+    builder.add(value);
+  }
+}
+
+arangodb::Result setObjectIdsForCollection(arangodb::LogicalCollection& collection,
+                                           std::pair<std::uint64_t, std::uint64_t>& outputPair,
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:183]:
```suggestion
                                           std::pair<std::uint64_t, std::uint64_t> const& outputPair,
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 8]
```diff
@@ -69,16 +73,499 @@
...
+  }
+  auto cleanup = arangodb::scopeGuard([oldProps, &db, &key]() -> void {
+    rocksdb::WriteOptions wo;
+    rocksdb::WriteBatch batch;
+    auto value = arangodb::RocksDBValue::Collection(oldProps);
+    batch.Put(arangodb::RocksDBColumnFamily::definitions(), key.string(), value.string());
+    db.Write(wo, &batch);
+  });
+
+  res = ::setObjectIdsForCollection(collection, outputPair, indicesMap);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:261]: If there is a fatal error/system halt while we are in `setObjectIdsForCollection`, then the scopeGuard will not be executed anymore. Does this leave things in a broken state, or will we be able to recover from this state?

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 9]
```diff
@@ -69,16 +73,499 @@
...
+  TRI_voc_rid_t revision = TRI_StringToRid(p, l, false);
+  {
+    arangodb::velocypack::ObjectBuilder guard(&builder);
+    for (auto pair : arangodb::velocypack::ObjectIterator(oldDocument)) {
+      builder.add(pair.key);
+      builder.add(pair.value);
+    }
+  }
+
+  arangodb::velocypack::Slice newDocument = builder.slice();
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:300]: We need to make sure here (and I am not sure if this is currently guaranteed) that the new documents that are created have their system attributes at the beginning.
Documents in the database must have `_key`, `_id` and `_rev` first, followed by `_from` and `_to` for edge collections. All other user-defined attributes must go afterwards.

I also don't understand why we do iterate over the old document's attributes just to store them unmodified in a new Builder. If we don't rewrite the document data, then the contents of `newDocument` and `oldDocument` should be identical. And we can then remove the whole iterate-over-all-attributes thing and reuse `oldDocument` as it is. This will also eliminate my concerns.

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 10]
```diff
@@ -69,16 +73,499 @@
...
+                            arangodb::RocksDBValue& buffer, rocksdb::Slice& value,
+                            rocksdb::Slice oldKey, arangodb::LocalDocumentId& newId) {
+  arangodb::velocypack::Slice indexedValues = arangodb::RocksDBKey::indexedVPack(oldKey);
+  if (ridx.unique()) {
+    key.constructUniqueVPackIndexValue(ridx.tempObjectId(), indexedValues);
+    buffer = arangodb::RocksDBValue::UniqueVPackIndexValue(newId);
+    value = {buffer.string().data(), buffer.string().size()};
+    return;
+  }
+  key.constructVPackIndexValue(ridx.tempObjectId(), indexedValues, newId);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:430]: if we get here, `value` is left unmodified. Is that intentional?

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 11]
```diff
@@ -69,16 +73,499 @@
...
+  }
+
+  arangodb::velocypack::Slice doc = arangodb::RocksDBValue::data(ps);
+  return arangodb::transaction::helpers::extractRevFromDocument(doc);
+}
+
+void rewritePrimaryIndexEntry(arangodb::RocksDBIndex& ridx, arangodb::RocksDBKey& key,
+                              arangodb::RocksDBValue& buffer, rocksdb::Slice& value,
+                              rocksdb::Slice oldKey, rocksdb::Slice oldValue,
+                              arangodb::LocalDocumentId& newId) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:406]:
```suggestion
                              arangodb::LocalDocumentId const& newId) {
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 12]
```diff
@@ -69,16 +73,499 @@
...
+  arangodb::velocypack::StringRef docKey = arangodb::RocksDBKey::primaryKey(oldKey);
+  TRI_voc_rid_t revision = arangodb::RocksDBValue::revisionId(oldValue);
+  key.constructPrimaryIndexValue(ridx.tempObjectId(), docKey);
+  buffer = arangodb::RocksDBValue::PrimaryIndexValue(newId, revision);
+  value = {buffer.string().data(), buffer.string().size()};
+}
+
+void rewriteEdgeIndexEntry(arangodb::RocksDBIndex& ridx, arangodb::RocksDBKey& key,
+                           rocksdb::Slice oldKey, arangodb::LocalDocumentId& newId) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:415]:
```suggestion
                           rocksdb::Slice oldKey, arangodb::LocalDocumentId const& newId) {
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 13]
```diff
@@ -69,16 +73,499 @@
...
+  arangodb::velocypack::StringRef vertexId = arangodb::RocksDBKey::vertexId(oldKey);
+  key.constructEdgeIndexValue(ridx.tempObjectId(), vertexId, newId);
+}
+
+void rewriteVPackIndexEntry(arangodb::RocksDBIndex& ridx, arangodb::RocksDBKey& key,
+                            arangodb::RocksDBValue& buffer, rocksdb::Slice& value,
+                            rocksdb::Slice oldKey, arangodb::LocalDocumentId& newId) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:422]:
```suggestion
                            rocksdb::Slice oldKey, arangodb::LocalDocumentId const& newId) {
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 14]
```diff
@@ -69,16 +73,499 @@
...
+    key.constructUniqueVPackIndexValue(ridx.tempObjectId(), indexedValues);
+    buffer = arangodb::RocksDBValue::UniqueVPackIndexValue(newId);
+    value = {buffer.string().data(), buffer.string().size()};
+    return;
+  }
+  key.constructVPackIndexValue(ridx.tempObjectId(), indexedValues, newId);
+}
+
+void rewriteFulltextIndexEntry(arangodb::RocksDBIndex& ridx, arangodb::RocksDBKey& key,
+                               rocksdb::Slice oldKey, arangodb::LocalDocumentId& newId) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:434]:
```suggestion
                               rocksdb::Slice oldKey, arangodb::LocalDocumentId const& newId) {
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 15]
```diff
@@ -69,16 +73,499 @@
...
+  arangodb::Result res;
+  arangodb::RocksDBKey key;
+
+  auto rcoll = static_cast<arangodb::RocksDBMetaCollection*>(collection.getPhysical());
+  TRI_ASSERT(rcoll);
+  auto& ridx = static_cast<arangodb::RocksDBIndex&>(index);
+
+  rocksdb::ReadOptions ro;
+  auto bounds = ridx.getBounds();
+  ro.prefix_same_as_start = true;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:509]:
```suggestion
  ro.prefix_same_as_start = index.type() != Index::TRI_IDX_TYPE_EDGE_INDEX;

```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 16]
```diff
@@ -69,16 +73,499 @@
...
+      return res;
+    }
+  }
+
+  std::vector<std::shared_ptr<arangodb::Index>> indices = rcoll.getIndexes();
+  for (auto& idx : indices) {
+    auto& ridx = static_cast<arangodb::RocksDBIndex&>(*idx);
+    if (ridx.tempObjectId() != 0) {
+      auto bounds = ridx.getBounds(ridx.tempObjectId());
+      res = arangodb::rocksutils::removeLargeRange(&db, bounds, true, true);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:556]:
```suggestion
      res = arangodb::rocksutils::removeLargeRange(&db, bounds, ridx.type() != Index::TRI_IDX_TYPE_EDGE_INDEX, true);
```

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 17]
```diff
@@ -1240,6 +1731,90 @@ void RocksDBCollection::adjustNumberDocuments(transaction::Methods& trx, int64_t
...
+  {
+    // start an exclusive transaction to block access to the collection
+    std::shared_ptr<transaction::Context> context =
+        transaction::StandaloneContext::Create(_logicalCollection.vocbase());
+    SingleCollectionTransaction trx(context, _logicalCollection, AccessMode::Type::EXCLUSIVE);
+    res = trx.begin();
+    if (res.fail()) {
+      return res;
+    }
+    auto trxGuard = scopeGuard([&trx]() -> void { trx.abort(); });
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1762]: I don't think we need this guard. Transactions are automatically aborted when they go out of scope and there hasn't been an explicit commit.

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 18]
```diff
@@ -1240,6 +1731,90 @@ void RocksDBCollection::adjustNumberDocuments(transaction::Methods& trx, int64_t
...
+
+    res = ::copyCollectionToNewObjectIdSpace(*engine.db(), _logicalCollection);
+    if (res.fail()) {
+      return res;
+    }
+
+    std::vector<std::shared_ptr<Index>> indices = getIndexes();
+    for (auto& index : indices) {
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1770]:
```suggestion
    for (auto const& index : indices) {
```

File: arangod/RocksDBEngine/RocksDBValue.h [hunk 19]
```diff
@@ -129,7 +129,7 @@ class RocksDBValue {
       : _type(other._type), _buffer(std::move(other._buffer)) {}
 
   RocksDBValue& operator=(RocksDBValue&& other) noexcept {
-    TRI_ASSERT(_type == other._type);
+    _type = other._type;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBValue.h:132]: Is this change still required? If not, I would like to revert it.

File: tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+      assertEqual(doc.num, i);
+      assertEqual(doc.num10, i % 10);
+      assertEqual(doc.location, { coordinates: [i, i], type: 'Point' });
+    }
+
+    // verify indexes
+    const indices = c.getIndexes();
+    assertEqual(indices.length, 5);
+
+    // primary
```
Comment (jsteemann) [inline tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js:59]: I think that upgrading edge indexes has an issue (please refer to the `prefix_same_as_start` comments above), so we should add tests here for an edge collection as well!

File: arangod/VocBase/Methods/Collections.cpp (hunk 4 above)
Comment (danielhlarkin) [reply arangod/VocBase/Methods/Collections.cpp:899]: Ah, no, you are correct. This is a mistake.

File: arangod/VocBase/LogicalCollection.cpp (hunk 5 above)
Comment (danielhlarkin) [reply arangod/VocBase/LogicalCollection.cpp:959]: This was a concern for me as well. I added this while I was working on the cluster upgrade procedure, as I didn't see a way around setting the properties in the cluster without using this method to react to plan changes or whatever. I can remove it for now and re-add if it's still necessary later on.

File: arangod/RocksDBEngine/RocksDBValue.h (hunk 19 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBValue.h:132]: Yes, it's needed to allow for holding a `RocksDBValue` buffer which is of unknown type at time of allocation. I'll add an assertion to ensure it is either assigning a value of the same type or that the initial type was `Placeholder`.

File: 3rdParty/fuerte/include/fuerte/message.h [hunk 21]
```diff
@@ -180,6 +180,8 @@ class Request final : public Message {
   Request(RequestHeader messageHeader = RequestHeader())
       : header(std::move(messageHeader)), _timeout(defaultTimeout) {}
 
+  Request(Request const&);
```
Comment (graetzer) [inline 3rdParty/fuerte/include/fuerte/message.h:183]: is this still necessary ?

File: tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js (hunk 20 above)
Comment (danielhlarkin) [reply tests/js/server/upgrade-data/upgrade-data-collection-upgrade-msv3.6.0.js:59]: I will add a test. Need to generate new data for the upgrade-data-tests repo first...

File: 3rdParty/fuerte/include/fuerte/message.h (hunk 21 above)
Comment (danielhlarkin) [reply 3rdParty/fuerte/include/fuerte/message.h:183]: Nope, will delete.

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp (hunk 6 above)
Comment (graetzer) [reply arangod/RocksDBEngine/RocksDBMetaCollection.cpp:637]: dito

Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBMetaCollection.cpp:637]: Good catch.

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 8 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCollection.cpp:261]: We should be able to handle it in recovery. This `scopeGuard` is just to fix the state if we encounter an error and need to still function without restarting the server.

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 9 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCollection.cpp:300]: Remnant from a temporary state. Can just copy the document as suggested.

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 10 above)
Comment (danielhlarkin) [reply arangod/RocksDBEngine/RocksDBCollection.cpp:430]: Yes, The non-unique index entry stores the revision in the key rather than the value.

File: arangod/Cluster/AgencyPaths.h [hunk 22]
```diff
@@ -1603,7 +1603,7 @@ class Root : public std::enable_shared_from_this<Root>, public Path {
 
           using BaseType::DynamicComponent;
 
-          class TransactionId : public StaticComponent<TransactionId, Server> {
```
Comment (jsteemann) [inline arangod/Cluster/AgencyPaths.h:1606]: these changes don't look intentional but rather like from a global search&replace.
the same comment applies for the following replacements of `TransactionId` with `TRI_voc_tid_t`.

File: arangod/Transaction/ClusterUtils.cpp [hunk 23]
```diff
@@ -40,7 +40,7 @@ void abortLeaderTransactionsOnShard(TRI_voc_cid_t cid) {
   TRI_ASSERT(mgr != nullptr);
   
   bool didWork = mgr->abortManagedTrx([cid](TransactionState const& state, std::string const& /*user*/) -> bool {
-    if (transaction::isLeaderTransactionId(state.id())) {
```
Comment (jsteemann) [inline arangod/Transaction/ClusterUtils.cpp:43]: these changes don't look intentional but rather like from a global search&replace.

File: tests/Cluster/AgencyPathsTest.cpp [hunk 24]
```diff
@@ -156,7 +156,7 @@ static_assert(!std::is_default_constructible<Root::Arango::Target::CleanedServer
 static_assert(!std::is_default_constructible<Root::Arango::Target::LatestCoordinatorId>::value, CONSTRUCTIBLE_MESSAGE);
 static_assert(!std::is_default_constructible<Root::Arango::Target::MapUniqueToShortId>::value, CONSTRUCTIBLE_MESSAGE);
 static_assert(!std::is_default_constructible<Root::Arango::Target::MapUniqueToShortId::Server>::value, CONSTRUCTIBLE_MESSAGE);
-static_assert(!std::is_default_constructible<Root::Arango::Target::MapUniqueToShortId::Server::TransactionId>::value, CONSTRUCTIBLE_MESSAGE);
```
Comment (jsteemann) [inline tests/Cluster/AgencyPathsTest.cpp:159]: these changes don't look intentional but rather like from a global search&replace.

File: arangod/Cluster/AgencyPaths.h (hunk 22 above)
Comment (danielhlarkin) [reply arangod/Cluster/AgencyPaths.h:1606]: Oh, shit, yeah that's my bad. Was messing around with a different checkout and must have accidentally misfired.

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/9380/

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/9385/
